src/coreclr/tools/Common/TypeSystem/IL/ILProvider.cs

```diff
@@ -11,6 +11,12 @@ namespace Internal.IL
     /// </summary>
     public abstract class ILProvider
     {
+        private int _version = 0;
+        protected void IncrementVersion()
+        {
+            _version++;
+        }
+        public int Version => _version;
         public abstract MethodIL GetMethodIL(MethodDesc method);
```

Comment:
Could you move this to `ReadyToRunILProvider` and retype a couple things to fix build breaks? There will be one place where hard cast is necessary in `UseILProvider(ILProvider ilProvider)`. I think that's fine.